packages/xo-vmdk-to-vhd/src/vmdk-read-table.js

```diff
@@ -51,47 +60,52 @@ export default async function readVmdkGrainTable(fileAccessor) {
-    true
-  )
-  const grainTablePhysicalSize = numGTEsPerGT * 4
-  const grainDirectoryEntries = Math.ceil(grainCount / numGTEsPerGT)
-  const grainDirectoryPhysicalSize = grainDirectoryEntries * 4
-  const grainDir = new Uint32Array(
-    await fileAccessor(
-      grainDirPosBytes,
-      grainDirPosBytes + grainDirectoryPhysicalSize
+  async function readTable() {
+    const grainSize =
+      getLongLong(headerBuffer, GRAIN_SIZE_OFFSET, 'grain size') * SECTOR_SIZE
+    const grainCount = Math.ceil(capacity / grainSize)
+    const numGTEsPerGT = new DataView(headerBuffer).getUint32(
+      NUM_GTE_PER_GT_OFFSET,
+      true
     )
-  )
-  const cachedGrainTables = []
-  for (let i = 0; i < grainDirectoryEntries; i++) {
-    const grainTableAddr = grainDir[i] * SECTOR_SIZE
-    if (grainTableAddr !== 0) {
-      cachedGrainTables[i] = new Uint32Array(
-        await fileAccessor(
-          grainTableAddr,
-          grainTableAddr + grainTablePhysicalSize
-        )
+    const grainTablePhysicalSize = numGTEsPerGT * 4
+    const grainDirectoryEntries = Math.ceil(grainCount / numGTEsPerGT)
+    const grainDirectoryPhysicalSize = grainDirectoryEntries * 4
+    const grainDir = new Uint32Array(
+      await fileAccessor(
+        grainDirPosBytes,
+        grainDirPosBytes + grainDirectoryPhysicalSize
       )
+    )
+    const cachedGrainTables = []
+    for (let i = 0; i < grainDirectoryEntries; i++) {
+      const grainTableAddr = grainDir[i] * SECTOR_SIZE
+      if (grainTableAddr !== 0) {
+        cachedGrainTables[i] = new Uint32Array(
+          await fileAccessor(
+            grainTableAddr,
+            grainTableAddr + grainTablePhysicalSize
+          )
+        )
+      }
     }
-  }
-  const extractedGrainTable = []
-  for (let i = 0; i < grainCount; i++) {
-    const directoryEntry = Math.floor(i / numGTEsPerGT)
-    const grainTable = cachedGrainTables[directoryEntry]
-    if (grainTable !== undefined) {
-      const grainAddr = grainTable[i % numGTEsPerGT]
-      if (grainAddr !== 0) {
-        extractedGrainTable.push([i, grainAddr])
+    const extractedGrainTable = []
+    for (let i = 0; i < grainCount; i++) {
+      const directoryEntry = Math.floor(i / numGTEsPerGT)
+      const grainTable = cachedGrainTables[directoryEntry]
+      if (grainTable !== undefined) {
+        const grainAddr = grainTable[i % numGTEsPerGT]
+        if (grainAddr !== 0) {
+          extractedGrainTable.push([i, grainAddr])
+        }
       }
     }
+    extractedGrainTable.sort(
+      ([i1, grainAddress1], [i2, grainAddress2]) =>
+        grainAddress1 - grainAddress2
+    )
+    return extractedGrainTable.map(([index, grainAddress]) => index * grainSize)
   }
-  extractedGrainTable.sort(
-    ([i1, grainAddress1], [i2, grainAddress2]) => grainAddress1 - grainAddress2
-  )
-  return extractedGrainTable.map(([index, grainAddress]) => index * grainSize)
+
+  return { tablePromise: readTable(), capacityBytes: capacity }
```

Comment:
Why did you choose to return a promise?

Reply:
the table extraction can be a long process, that way it start immediately upon dropping the file on the UI, but it only becomes blocking when we submit the file.

Reply:
This is weird, but as long as it works.